tests/unit/expression_test.py

```diff
@@ -2453,3 +2457,164 @@ def test_proof_invalid_logic() -> None:
+
+    assert_type_error(
+        CaseExpr(c1, [([ID("V1"), ID("V2")], TRUE), ([ID("V3")], Number(1))]),
+        r'^model: error: dependent expression "True" has incompatible enumeration type '
+        r'"__BUILTINS__::Boolean"\n'
+        r'model: info: conflicting with "1" which has type universal integer \(1\)$',
+    )
+    assert_type_error(
+        CaseExpr(Opaque(Variable("X", type_=rty.Message("A"))), [([ID("V")], Number(1))]),
+        r'^model: error: invalid discrete choice with sequence type "__INTERNAL__::Opaque" '
+        r'with element integer type "Byte" \(0 .. 255\)\n'
+        r"model: info: expected enumeration or integer type$",
+    )
+
+
+def test_case_simplified() -> None:
+    assert_equal(
+        CaseExpr(
+            Variable("C"), [([ID("V1"), ID("V2")], And(TRUE, FALSE)), ([ID("V3")], FALSE)]
+        ).simplified(),
+        CaseExpr(Variable("C"), [([ID("V1"), ID("V2")], FALSE), ([ID("V3")], FALSE)]),
+    )
+
+
+def test_case_invalid() -> None:
+    assert_type_error(
+        CaseExpr(
+            Variable("C", type_=ENUMERATION.type_),
+            [([ID("Zero")], TRUE), ([ID("One")], FALSE)],
+            location=Location((1, 2)),
+        ),
+        "^<stdin>:1:2: model: error: not all enumeration literals covered by case expression\n"
+        '<stdin>:10:2: model: info: missing literal "Two"$',
+    )
+    assert_type_error(
+        CaseExpr(
+            Variable("C", type_=ENUMERATION.type_),
+            [([ID("Zero"), ID("One")], TRUE), ([ID("Two"), ID("Invalid")], FALSE)],
+            location=Location((1, 2)),
+        ),
+        "^<stdin>:1:2: model: error: invalid literals used in case expression\n"
+        '<stdin>:10:2: model: info: literal "Invalid" not part of "P::Enumeration"$',
+    )
+    assert_type_error(
+        CaseExpr(
+            Variable("C", type_=ENUMERATION.type_),
+            [
+                ([ID("Zero"), ID("One", location=Location((2, 2)))], TRUE),
+                ([ID("Two"), ID("One", location=Location((3, 2)))], FALSE),
+            ],
+            location=Location((1, 2)),
+        ),
+        "^<stdin>:1:2: model: error: duplicate literals used in case expression\n"
+        '<stdin>:3:2: model: info: duplicate literal "One"$',
+    )
+
+    assert_type_error(
+        CaseExpr(
+            Variable("C", type_=TINY_INT.type_),
+            [([Number(1)], TRUE), ([Number(2)], FALSE)],
+            location=Location((2, 2)),
+        ),
+        '^<stdin>:2:2: model: error: case expression does not cover full range of "P::Tiny"\n'
+        "<stdin>:1:2: model: info: missing value 3$",
+    )
+    assert_type_error(
+        CaseExpr(
+            Variable("C", type_=TINY_INT.type_),
+            [([Number(1)], TRUE)],
+            location=Location((2, 2)),
+        ),
+        '^<stdin>:2:2: model: error: case expression does not cover full range of "P::Tiny"\n'
+        "<stdin>:1:2: model: info: missing range 2-3$",
```

Comment:
How about using a more Ada-like syntax for the value ranges (i.e., `2 .. 3` instead of `2-3`)? Also, I think it would be good to have a bit more complex test case which also covers the "missing value X" case.

Reply:
I changed the syntax and also added a more complex test case involving ranges and single values.